library/core/src/char/decode.rs

```diff
@@ -91,7 +91,7 @@ impl<I: Iterator<Item = u16>> Iterator for DecodeUtf16<I> {
             None => self.iter.next()?,
         };
 
-        if u < 0xD800 || 0xDFFF < u {
+        if !u.is_char_surrogate() {
```

Comment:
Hmm, this makes me think that the API might want to be `-> Option<bool>` or something to distinguish high/low surrogates.

This code looks like it would be better written with exhaustive range patterns today (they almost certainly didn't exist when it was written, though).  So the nice code change, to me, would be one that could `match c.tell_me_about_surrogate_ness()` instead.